webui/ec2/EC2.py

```diff
@@ -23,78 +23,120 @@ def _user_add_keys(user):
             k.create_date = key.get('CreateDate')
             k.key_id = key.get('AccessKeyId')
             k.status = key.get('Status').lower()
-            user.keys.append(k);
+            user.keys.append(k)
     return user
 
 
-def get_users(name = None):
+def get_users(name=None):
     '''
-    See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html
-    for more details on boto iam api.
+    See for more details on boto iam api.
+        https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html
     '''
     iam = boto3.client('iam', aws_access_key_id=provider_conf.EC2['key'],
                        aws_secret_access_key=provider_conf.EC2['secret'])
     users = []
     if (name):
         try:
-            response = iam.get_user(UserName = name)
+            response = iam.get_user(UserName=name)
             user = response.get('User')
-            u = _ec2user_to_user(user);
-            _user_add_keys(u);
-            users.append(u);
+            u = _ec2user_to_user(user)
+            _user_add_keys(u)
+            users.append(u)
         except ClientError as ex:
-            print (ex);
+            print(ex)
     else:
         for response in iam.get_paginator('list_users').paginate():
             for user in response.get('Users'):
-                u = _ec2user_to_user(user);
-                _user_add_keys(u);
-                users.append(u);
+                u = _ec2user_to_user(user)
+                _user_add_keys(u)
+                users.append(u)
+
+    return users
+
+
+def get_user_by_key(key_id):
+    users = get_users()
 
-    return users;
+    for user in users:
+        for key in user.keys:
+            if key_id == key.key_id:
+                return user
+    return None
+
+
+def get_key(key_id):
+    users = get_users()
+    for user in users:
+        for key in user.keys:
+            if key_id == key.key_id:
+                return key
+    return None
+
+
+def delete_key(key_id):
+    user = get_user_by_key(key_id)
+    if (user is None):
+        return False
+    if (len(user.keys) == 1):
+        delete_user(user.name)
+        return True
+    else:
+        iam_res = boto3.resource(
+            'iam', aws_access_key_id=provider_conf.EC2['key'],
+            aws_secret_access_key=provider_conf.EC2['secret'])
+        key = iam_res.AccessKey(user.name, key_id)
+        key.delete()
+        return True
 
 
 def delete_user(username):
     iam = boto3.client('iam', aws_access_key_id=provider_conf.EC2['key'],
                        aws_secret_access_key=provider_conf.EC2['secret'])
     iam_res = boto3.resource('iam', aws_access_key_id=provider_conf.EC2['key'],
-                       aws_secret_access_key=provider_conf.EC2['secret'])
+                             aws_secret_access_key=provider_conf.EC2['secret'])
 
-    user = iam_res.User(username)
-    user.load()
+    try:
+        user = iam_res.User(username)
```

Comment:
there is https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html#IAM.Client.get_user function , any good reason to use User and catch exception instead of calling get user and check for empty output ?

Reply:
main concern here that I haven't found clear list of cases when this ClientError might popup. Maybe it popup up with a lot of different reasons and you simplify it to "user not exists" state this may lead to unexpected behavior

Reply:
Don't understand in which direction you are thinking.
From my point, if the user doesn't exists. In case of unable to connect to Amazon or if it really not exists, this doesn't matter, cause in both cases I should stop the request.
Do you expect different exit codes in our REST API in such case?

Reply:
I overlook that you actually using this `user` object later so all is fine here